src/country_list.rs

```diff
@@ -0,0 +1,203 @@
+use gtk::prelude::*;
+use gtk::subclass::prelude::*;
+use gtk::{gio, glib};
+use indexmap::IndexMap;
+use std::iter::FromIterator;
+use tdlib::types;
+
+use crate::country_info::CountryInfo;
+
+mod imp {
+    use super::*;
+
+    use once_cell::unsync::OnceCell;
+
+    #[derive(Debug, Default)]
+    pub(crate) struct CountryList {
+        pub(super) list: OnceCell<IndexMap<String, CountryInfo>>,
+        pub(super) invalid: CountryInfo,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for CountryList {
+        const NAME: &'static str = "CountryList";
+        type Type = super::CountryList;
+        type ParentType = glib::Object;
+        type Interfaces = (gio::ListModel,);
+    }
+
+    impl ObjectImpl for CountryList {}
+
+    impl ListModelImpl for CountryList {
+        fn item_type(&self, _list_model: &Self::Type) -> glib::Type {
+            CountryInfo::static_type()
+        }
+
+        fn n_items(&self, list_model: &Self::Type) -> u32 {
+            list_model.list().len() as u32 + 1
+        }
+
+        fn item(&self, list_model: &Self::Type, position: u32) -> Option<glib::Object> {
+            let list = list_model.list();
+            if position as usize == list.len() {
+                // The last item in the list is always the invalid `CountryInfo`.
```

Comment:
It's a bit weird to have an "invalid" item in the list model, this seems something that should be handled in the PhoneNumberInput widget instead, probably with an expression that can check if the selected item is invalid and show the "invalid" (or "unknown") text in case it is.

Reply:
Yes agree. Not the best idea to let the user choose "invalid" from the list. I removed it. I have not found a way to make `CombRow` display custom text. So, now the subtitle of the row will have an error hint and nothing is selected in the combo box if the country code the user enters is invalid.

Reply:
Ah, apparently the ComboRow::expression doesn't work if the selected item is null, so my suggestion doesn't work. So, I think your solution is ok.